### ISSUE TITLE
Create port mapping using STUN protocol

### ISSUE THREAD

**Issue by mnpp** [miniupnp/miniupnp#322]
This is a feature request to add support for miniupnpd to process STUN requests to create port mappings.

When miniupnpd receives a STUN "Binding" request, it could create a port mapping and return the public IP:port in the "Binding" response. This would be similar to a STUN server, the difference is that the service would run on the LAN interface serving LAN clients only. It effectively turns a STUN-like service into a port mapping service like PCP/NAT-PMP/UPnP.

This would be useful to people who use SIP VoIP clients because many clients (both software and hardware) only support STUN which does not work with certain types of NAT (e.g. symmetric NAT), these NAT problems can be solved by creating port mappings mentioned above.

STUN standard is defined in RFC 5389, which should be enough to create port mappings.

Optionally there is also an experimental standard RFC 5780 which adds support for "NAT Behavior Discovery", that requires the server to have 2 IP addresses and 2 ports, I think 2 private LAN IPs should work which is relatively easy to configure on a router running custom firmware/software like OpenWrt

**Comment by pali** [comment]
Such thing which you described here would never work and does not make sense at all.

Reasons:

1. STUN binding request does not contain any information for port forwarding request. So there is absulutely no way for STUN server (or any other proxy or MITM-proxy) to retrieve information what is the port on which STUN client want to receive data from external internet server.

2. It cannot be used and therefore cannot be useful for SIP VoIP clients. VoIP clients use STUN protocol for determining external IPv4 address and type of NAT. They are putting public IPv4 address (obtained by STUN) into SDP packets (for RTP) or into SIP methods instead of IPv4 address of local interface. Some SIP registrars refuse to REGISER client if it does not "know" public IPv4 address. When SIP client detects symmetric NAT, then it sometimes (or more times) does not do any changes and send SDP and SIP packets with private IPv4 addresses. Lot of SIP servers (Asterisk, Kamailio, ...) have NAT detection plugins (based on presence of private IPv4 addresses in SIP/SDP packets) and automatically enable own RTP relay and put into SIP/SDP packets relay IPv4 address.

3. It cannot be useful when you are behind symmetric NAT. In such case miniupnpd cannot work, so it cannot do anything with clients which are "behind" miniupnpd. In all other cases SIP VoIP clients would work like with external STUN server, so having something like STUN server in miniupnpd does not bring anything.

4. STUN server implemented by RFC 5389 does not provide anything which can be used to create port mappings for port forwarding. None of provided attributes defined in RFC 5389 (https://tools.ietf.org/html/rfc5389#section-15) can be used for it.

The only possible way how to use STUN protocol for port mapping is to extend STUN protocol and use some reserved attribute (or register new STUN attribute in IANA) for sending port to STUN server. But such thing is not supported by any existing client, therefore usefullness of it would be zero. List of all registered attributes in IANA are on: https://www.iana.org/assignments/stun-parameters/stun-parameters.xhtml

**Comment by mnpp** [comment]
This request isn't about implementing a STUN server in miniupnpd or extending the STUN protocol. It's about using the existing STUN protocol to create port mappings (which is not something a STUN server can do), that's why I described it as a STUN-like service, but from the client's perspective, it's STUN-compliant (except that it's running on LAN interface).

A STUN client uses a binding request to discover the public IP address and port. A typical usage (e.g. SIP VoIP client) is that after learning the public IP:port, the client would inform the remote endpoint to send data to the public IP:port. e.g. client sends STUN request from 10.0.0.2:1234, STUN server sees the request coming from 1.1.1.1:6789, so it returns 1.1.1.1:6789 to the client, then client tells remote endpoint to send data to 1.1.1.1:6789. The binding request can be treated as a port mapping request, the client asks the server to map an arbitrary external port to the client's source port that's used to send the request and tell the client what that external port is in the binding response.

Using the example above, the proposed feature is for miniupnpd to process the STUN request coming from 10.0.0.2:1234, and create a port mapping from an arbitrary external port (although port 1234 or something close to it would be preferred) to 10.0.0.2:1234 and return the public IP:port to the client, so that data sent from any remote endpoint to the external port could be received by the client. Without this port mapping, data sent from a different port and/or IP may not reach 10.0.0.2:1234 depending on NAT implementation.

The experimental standard RFC 5780 could be optionally implemented to support "NAT Behavior Discovery", it's not about creating port mappings. If port mapping mentioned above is created successfully, and RFC 5780 is implemented, miniupnpd could return responses to indicate that it's a Full Cone NAT or least restrictive NAT.

**Comment by pali** [comment]
> client sends STUN request from 10.0.0.2:1234, STUN server sees the request coming from 1.1.1.1:6789, so it returns 1.1.1.1:6789 to the client

Yes.

> then client tells remote endpoint to send data to 1.1.1.1:6789.

Not exactly.

Client may tell remote endpoint to send data to 1.1.1.1:6789 only in the one case: when NAT forward traffic from *any* remote address which comes to the port 6789 to the 10.0.0.2:1234. And this happen only when NAT is configured with static port forwarding already. Normally NAT forward this traffic only from STUN server address (to which client send STUN binding request).

What you described here is in STUN protocol defined as Full Cone NAT, but it is not widely used. I have not seen applications which try to work with this strategy as it is very fragile. NAT implemented/configured by Linux netfiler with conntrack use strategy described in previous paragraph (I hope it matches Port Restricted Cone NAT definition).

**Comment by mnpp** [comment]
> Client may tell remote endpoint to send data to 1.1.1.1:6789 only in the one case: when NAT forward traffic from any remote address which comes to the port 6789 to the 10.0.0.2:1234

This depends on client implementation, some clients may use the discovered public IP:port, some may not, but I have seen applications that do use the discovered public IP:port, and they still work most of the time because these clients would then send data from the same private IP:port to the remote endpoint which would cause (some) NAT to create the necessary mapping to forward incoming data to the client, however if the  client does not send data to the remote endpoint using the same private port, or a different external port is used by NAT (e.g. symmetric NAT), it wouldn't work. So it is fragile, that's one of the reasons why I raised this feature request - if a port mapping is created in this process (and ideally implement RFC 5780 to indicate it's a Full Cone NAT), the problem would go away - it would satisfiy the condition you described:
> when NAT forward traffic from any remote address which comes to the port 6789 to the 10.0.0.2:1234

Additionally it would even work with symmetric NAT (as it would cease being a symmetric NAT after port mapping is created).

**Comment by mnpp** [comment]
Example:

A STUN binding request from 10.0.0.2:1234 can be treated and processed as a PCP MAP Opcode Request with the following attributes:
```
Protocol = 17 (UDP)
Internal Port = 1234
Suggested External Port = 1234
...
```

A successful PCP MAP Opcode Response to the request above can be returned to the client in a STUN binding response with the following attributes:
```
Binding Response:Port = PCP MAP Opcode Response:Assigned External Port
Binding Response:Address = PCP MAP Opcode Response:Assigned External IP
Binding Response:X-Port = XOR (PCP MAP Opcode Response:Assigned External Port)
Binding Response:X-Address = XOR (PCP MAP Opcode Response:Assigned External IP)
...
```

**Comment by pali** [comment]
STUN binding request is sent by client from randomly chosen local port. SIP (UDP) port is 5060 and STUN clients does not have to sent binding request from port 5060. And most of them does not do that. Also RTP ports in SIP are in the most implementations randomly chosen.

If you found implementation of SIP/SDP/RTP which use one and the same UDP port for everything, including STUN bind request, it does not mean that it is normal, nor that such thing is standard.

For the most SIP/VOIP clients your suggested idea would not work and miniupnpd is not able to detect if STUN client which sent binding request is one of those which you described (use one UDP port for everything).

Therefore for the most SIP/VOIP clients this may be harmfull as it adds new and new mapping and forwarding rules which are fully useless.

And such thing even make fully compliant STUN client misbehave. Because symmetric or restricted NAT with this your suggested implementation would be detected by as full cone NAT by properly implemented STUN client. This is really dangerous and could make applications designed to work behind NAT stop working properly...

**Comment by pali** [comment]
That misbehave and misdetection would be caused by adding those new mapping rules for port forwarding. But because local port for binding request is randomly chosen and properly implemented UDP/SIP/whatever client would not use it for anything else, it would not be possible to detected that is behind restricted or symmetric NAT. And stops working correctly.

**Comment by mnpp** [comment]
A STUN client does not choose a port to send STUN request and then forget about that port unless the client only wants to detect its public IP address.

I'm not implying that a SIP VoIP client would only use one port for everything. Normally SIP VoIP clients use at least 2 ports, one for SIP signalling, one for RTP media. The SIP signalling port does not need to be 5060 unless you want to make direct calls between 2 UAs without specifying the singalling port. Normally calls are made through SIP server/proxy(s), it's the SIP server/proxy that normally uses 5060 as the signalling port. Normally it works like this
```
SIP.UA1.portX <-> SIP.server1.port5060 <-> SIP.server2.port5060 <-> SIP.UA2.portY
```
If UAs were required to use port 5060, then it would be impossible to use multiple UAs behind the same NAT.

Correct me if I'm wrong, presumably you think STUN is used in such way that a client first picks a random port to send STUN request and discovers the public IP:port and NAT type, then assume all ports would behave the same way. E.g. if the internal port is mapped to the same external port, it can then use another port to send/receive application data with the assumption that the new port would map to the same external port. That assumption may not always hold because e.g. the client may later coincidentally pick a port that's already mapped to another host behind NAT. Therefore it's important that the port used by the application to receive data is the same port used to send STUN request. RFC5780 also mentions in https://tools.ietf.org/html/rfc5780#section-4.1 that "Because mapping behavior can vary on a port-by-port basis, an application should perform its tests using the source port intended for use by the application whenever possible.  If it intends to use multiple source ports, it should repeat these tests for each source port."

It's normal for a SIP VoIP client to make 2 STUN requests, one for SIP signalling and one for RTP media, if this feature is implemented, two port forwarding rules would be created. With respect to these 2 ports, the NAT is Full Cone since traffic from any IP to these ports would be forwarded to the client behind NAT. This would work with fully compliant STUN clients. If a fully compliant STUN client needed to use more ports, the client shouldn't assume the NAT behaviour would be same as what it discovered earlier, more STUN requests would need to be sent due to the potential port clash mentioned above and the various scenarios in RFC5780.

For STUN clients that are compliant with RFC5780, the proposed feature shouldn't cause problems, if a user does not want the feature, I suppose there would be a flag to enable/disable this feature?

**Comment by pali** [comment]
> A STUN client does not choose a port to send STUN request and then forget about that port unless the client only wants to detect its public IP address.
> 
> I'm not implying that a SIP VoIP client would only use one port for everything. Normally SIP VoIP clients use at least 2 ports, one for SIP signalling, one for RTP media. The SIP signalling port does not need to be 5060 unless you want to make direct calls between 2 UAs without specifying the singalling port. Normally calls are made through SIP server/proxy(s), it's the SIP server/proxy that normally uses 5060 as the signalling port. Normally it works like this
> 
> ```
> SIP.UA1.portX <-> SIP.server1.port5060 <-> SIP.server2.port5060 <-> SIP.UA2.portY
> ```
> 
> If UAs were required to use port 5060, then it would be impossible to use multiple UAs behind the same NAT.

In SIP for signalling is used static port 5060. Lot of clients (hw or sw) allows you to change it, but default is 5060. SIP proxies are now smart enough and can handle connections also from different ports, e.g. when local port 5060 is translated to different by NAT.

What some SIP proxies/registers requires is to put correct public IPv4 address into SIP REGISTER method. And for this reason client sends its first STUN binding request. Not for port mapping, but for detection of public IPv4 address. Plus it may detect restrictive NAT (e.g. symmetric) and decide to use later some relay instead of trying direct connections. So for these binding requests you really do not want to create port forwarding rules.

In next steps, when doing INVITE, you need two UDP ports for RTP (but in most situations, one is enough; second is for RTP control). Here you need to know exact public IPv4 address and port number to send it via SIP SDP to other side. But now when SIP proxies have support for relays and automatically rewrite SDP packets based on private/local IPv4 addresses, SIP clients try to put public IPv4 addresses+ports only if they are sure that NAT traversal would work -- e.g. when they are not behind symmetric NAT.

So when symmetric NAT is used, mapping and port forwarding rules are useless there -- they would not work correctly. For other types of NAT it may be useful, but... SIP clients already handle this situation in better way: Both RTP sides start sending UDP packets to open NAT incoming ports by NAT traversal technique.

> Correct me if I'm wrong, presumably you think STUN is used in such way that a client first picks a random port to send STUN request and discovers the public IP:port and NAT type, then assume all ports would behave the same way.

This is used for detection of restrictive NAT type (like symmetric). If you got response from STUN with different port (or no answer when you ask STUN server to use different address/port), then you know that there is some restrictive NAT which cause problems for direct connections. And you should use some relay. If at least one port is filtered or translated, you cannot assume that other ports would not be.

> For STUN clients that are compliant with RFC5780, the proposed feature shouldn't cause problems, if a user does not want the feature, I suppose there would be a flag to enable/disable this feature?

Flag to enable/disable could solve this problem. But I'm really skeptical for benefits. There are lot of software which mangle SIP/SDP packets (including Linux kernel netfilter/iptables modules) which just cause problems or make SIP software misbehave. If somebody has problems with SIP client, first thing is to disable everything which mangle/change/do-SIP-NAT or whatever. As today either SIP proxies support relaying or SIP clients are smart enough to handle NATs. And this feature is another thing which is going to change behavior how SIP packets are going routed/filtered/NATed by Linux kernel.

**Comment by mnpp** [comment]
The point I was making about UA signalling port is that it doesn't have to be 5060 although 5060 is the default, therefore there shouldn't be problems sending a binding request from a non-default signalling port or forwarding a non-default signalling port.

How a SIP client uses STUN for the REGISTER request is again dependent on implementation, some implementation may just use the discovered IP address as you said, some may use both the discovered IP address and port and some may do something else. Here is one example using both the IP and port https://www.3cx.com/blog/voip-howto/stun-protocol/ . However, I'm less concerned about the use of STUN for REGISTER as I normally don't have problems with REGISTER with or without STUN.

The most common SIP VoIP problem is one-way audio. There're different solutions that are supposed to address this problem, client side, server side and NAT gateway. Usually NAT gateway solutions such as NAT SIP helper and SIP ALG etc are the worse, they tend to break things than to fix them. I always disable them regardless whether there is a problem. That leaves me with 2 choice, client side or server side (or both which I would avoid). There're SIP providers that have a decent server side solution, but some providers do not, and sometimes I need to use these providers, so I have to find a client side solution. Another benefit of client side solution is that it may allow media stream to flow directly between endpoints bypassing the server, this would reduce the latency.

STUN is a client side solution, but it alone may not work with some network configurations, e.g. symmetric NAT, but with the proposed feature, it would work because once the RTP port is forwarded, NAT type for that port would change from symmetric to full cone. I don't know why you think it wouldn't work - with this feature, a STUN binding response containing the forwarded external port would be sent to the client and client would inform the remote endpoint to send data to that port.

> This is used for detection of restrictive NAT type (like symmetric). If you got response from STUN with different port (or no answer when you ask STUN server to use different address/port), then you know that there is some restrictive NAT which cause problems for direct connections. And you should use some relay. If at least one port is filtered or translated, you cannot assume that other ports would not be.

Neither can you assume that other ports would always be. I treat NAT type as a port-dependent property because in https://tools.ietf.org/html/rfc5780#section-4  "Results from tests performed using a particular port on the client may also not indicate the behavior experienced by a different port"

That's why https://tools.ietf.org/html/rfc5780#section-4.1 recommends that "Because mapping behavior can vary on a port-by-port basis, an application should perform its tests using the source port intended for use by the application whenever possible." Therefore if an application knows what port it's going to use later that requires NAT traversal, it should use that port to send STUN requests instead of using another random port. So it's reasonable to assume that the source port in the binding request is a port the client intends to use (assuming the client follows RFC5780 recommendation), and with the proposed feature, a port forwarding rule could be created in response to the binding request to make NAT less restrictive for that port making the application more likely to work.

**Comment by pali** [comment]
> STUN is a client side solution, but it alone may not work with some network configurations, e.g. symmetric NAT, but with the proposed feature, it would work because once the RTP port is forwarded, NAT type for that port would change from symmetric to full cone. I don't know why you think it wouldn't work - with this feature, a STUN binding response containing the forwarded external port would be sent to the client and client would inform the remote endpoint to send data to that port.

miniupnpd does not work behind symmetric NAT (nor any other restrictive NAT). Therefore it does not make sense to create a mapping rules in miniupnpd which itself is not able to forward ports because is behind symmetric NAT.

And if you are talking that routing BOX on which is running some kind of NAT and miniupnpd too, why you want to create something for miniupnpd to use STUN for local clients, if easier and more powerful solution is to change NAT from symmetric to either full cone or port/address restricted? If you already have access to that routing BOX (because you are going to update/patch miniupnpd for having port forwarding based on client STUN binding requests), you can rather change NAT type and you do not need such things like port forwarding based on STUN. I see that in some situations it may work, but why? This is really anti-pattern solution which may misbehave some clients.

And what you described here, is lot of easier to implement via existing netfilter infrastructure. IIRC it should be already possible via iptables and ipset rules, without any patching. Match outgoing UDP packet which is going to destination STUN port, store its source UDP port into ipset for some period and then accept all UDP packets which goes to the UDP port stored in ipset. Maybe there is needed some route tweaking. And if it is not easily possible (because I do not see something...) still it better match usage of conntrack helper (either userspace or kernel) and not miniupnpd.

I guess this is simple, faster (as implemented in kernel netfilter), plus more systematic solution.

**Comment by mnpp** [comment]
This is about running miniupnpd on the router that performs NAT, not about running miniupnpd behind NAT.

I could patch the NAT code or create firewall rules to fix the problem, but that's inconvenient, it's the same reason why people use miniupnpd instead of manually creating firewall rules.

From a client's perspective, this solution is STUN compliant, the feature only makes NAT less restrictive, it's not an anti-pattern, the process is equivalent to a person manually forwarding a port followed by a STUN client testing that port, if a STUN client works in an unexpected way due to port forwarding, then it's the client that's not STUN compliant, the client shouldn't care about whether the tested port is forwarded, it should just process the STUN response and react accordingly. This feature simply automates that process and generates a STUN response that reflects the true state of NAT for the tested port which would be full cone.

**Comment by pali** [comment]
> I could patch the NAT code or create firewall rules to fix the problem, but that's inconvenient

Why?

> it's the same reason why people use miniupnpd instead of manually creating firewall rules.

It is not, it is something totally different. PCP or IGD UPnP are protocol specially designed for configuring port forwarding. STUN was never designed for port forwarding rules and also it does not contain any attributes for them inside. You just trying to misuse it based on source/target port addresses and how it works.

Yea, it may work in some situation, but whole design like you described better match conntrack or iptables. It is really anti-pattern to have something in miniupnpd.

Why not write iptables/ipset rules or (if it not possible) extend conntrack helper to implement this functionality? What you described here is exactly what conntrack is doing, so it better fits into conntrack as into miniupnpd.

**Comment by pali** [comment]
>  find patching and compiling the kernel whenever I need to upgrade the kernel inconvenient.

Sorry, but I do not understand now. Why patching and compiling kernel? I have not suggested nor wrote nothing like that. You just need to have kernel compiled with iptables supports. But if you have kernel without netfilter, then you need to recompile it even for usage of miniupnpd... as miniupnpd does not work without netfilter.

> This is not misusing STUN protocol, it's about performing additional action which miniupnpd is capable of in response to a STUN request. As mentioned earlier, it's equivalent to a person manually forwarding a port right before a client makes a STUN request, how is that misusing STUN and an anti-pattern?

Is STUN protocol designed and defined to do such operation? No, it is not and so it is misusage of that protocol.

What you described here is really connection tracking based on STUN binding request packets. And such thing really should be handled at proper layer -- conntrack.

miniupnpd handles explicit requests for port forwarding by protocols which have information about port forwarding request and are designed for them. I really do not think that port forwarding should be done by STUN binding request packets.

**Comment by mnpp** [comment]
> Sorry, but I do not understand now. Why patching and compiling kernel? I have not suggested nor wrote nothing like that. You just need to have kernel compiled with iptables supports. But if you have kernel without netfilter, then you need to recompile it even for usage of miniupnpd... as miniupnpd does not work without netfilter.

I said earlier I could patch NAT code or create firewall rules to fix the problem. The former would require kernel patching and compiling.

The proposed feature would perform additional action in response to STUN request, I wouldn't call that misuse, otherwise all NAT helpers would be misusing all sorts of protocols like FTP, PPTP and SIP etc because the helpers do things not defined in these protocols.

The proposed feature can be treated like a STUN-helper. It can be implemented in different layers in the same way that there're NAT SIP helper in the network layer, server and client side NAT traversal solutions in the application layer. Being able to implement it in the network layer should not prevent it from being implemented in the application layer, besides doing it in the network layer would impact the performance a little since packets would need to be inspected to see if it's STUN. In my opinion, miniupnpd is a good candidate to implement this feature because it already supports port-forwarding which is the the most important function required by this feature.

There are situations where port-forwarding isn't needed when a STUN binding request is made, users would have the option to not use miniupnpd as the STUN server in the client or run miniupnpd with this feature disabled. I think it's better to give users the choice than to not have the feature at all.

**Comment by pali** [comment]
> I said earlier I could patch NAT code or create firewall rules to fix the problem.

So why not to prepare rules and then nothing is needed to implement in miniupnpd?

> The former would require kernel patching and compiling.

Why? Adding new iptables/ipset rules do not need kernel patching nor compiling -- in case you already have kernel with netfilter support (which probably all major distributions have).

**Comment by pali** [comment]
> The proposed feature would perform additional action in response to STUN request, I wouldn't call that misuse, otherwise all NAT helpers would be misusing all sorts of protocols like FTP, PPTP and SIP etc because the helpers do things not defined in these protocols.

Well, those are protocol specific rules and you exactly know what commands and data traffic is expected. But STUN protocol itself does not say what data, from what port or address is expected.

> The proposed feature can be treated like a STUN-helper. It can be implemented in different layers in the same way that there're NAT SIP helper in the network layer, server and client side NAT traversal solutions in the application layer. Being able to implement it in the network layer should not prevent it from being implemented in the application layer, besides doing it in the network layer would impact the performance a little since packets would need to be inspected to see if it's STUN.

> In my opinion, miniupnpd is a good candidate to implement this feature because it already supports port-forwarding which is the the most important function required by this feature.

But port-forwarding is something different as you described in above SIP scenario. You want to track RTP ports together with STUN binding request. Your SIP client does not specify that want to forward port (from public to private) for specified amount of time. But rather your SIP client want to have active communication for active call time. Also via STUN it cannot inform miniupnpd that call was finished and forwarding entry should be cleaned. Instead RTP connection should be tracked and closed by NAT box immediately after call finish on inactivity. Port forwarding in miniupnpd is not build on connection traffic, but on timers.

I'm still in doubt that this would ever work properly... For how long period would you put port-forwarding entry into iptables? How you would know that active voice call would not be longer then time period used by miniupnpd?

> There are situations where port-forwarding isn't needed when a STUN binding request is made, users would have the option to not use miniupnpd as the STUN server in the client or run miniupnpd with this feature disabled. I think it's better to give users the choice than to not have the feature at all.

And how you would describe users when they need such thing and when not? For me it is still not obvious and based on above timing problems, I'm still skeptical that it may work reliable.

If you put port forwarding entry into miniupnpd for infinite time, then port forwarding table just start to be huge and after some time you eat all 2^16 usable ports. And when you put it for less then infinite time you would have a problem for voice calls which are longer then what you specified.

**Comment by miniupnp** [comment]
https://people.netfilter.org/chentschel/docs/sip-conntrack-nat.html

**Comment by mnpp** [comment]
I think you missed my point. Earlier I said "I could patch the NAT code or create firewall rules to fix the problem, but that's inconvenient, it's the same reason why people use miniupnpd instead of manually creating firewall rules."

"Patching NAT code" is patching the kernel. Alternatively firewall rules could be created every time client IP:port changes or new clients are added. Neither is convenient.

>So why not to prepare rules and then nothing is needed to implement in miniupnpd?

As mentioned earlier, it's due to laziness... It's really the same reason why miniupnpd exists, people don't have to use miniupnpd as everything miniupnpd does can also be done manually instead.

>Well, those are protocol specific rules and you exactly know what commands and data traffic is expected. But STUN protocol itself does not say what data, from what port or address is expected.

The source port of the STUN request is important if you read the RFC carefully.

How SIP/RTP works shouldn't matter here, this feature is not application specific, it's not about tracking connections of the applications that use STUN. It's about performing actions in response to STUN requests, so that applications using STUN would benefit and be more likely to work. If STUN request is processed in the way described earlier, applications using STUN such as SIP client would just work if they follow the STUN RFC recommendation.

Regarding the lifetime of the port-forwarding rules, it's a minor issue. NAT itself has a similar characteristic that applications cannot remove UDP translations or set the translation expiry, but we normally don't have problems with that. It shouldn't be difficult to deal with port-forwarding rule expiry. One solution is rely on the existing option "clean_ruleset_interval"

>https://people.netfilter.org/chentschel/docs/sip-conntrack-nat.html

I've tried that before, it's unreliable and sometimes caused problems. That's why I always disable NAT SIP helper and SIP ALG.

I think the proof of concept should be clear now. I don't think I have anything more to add.

**Comment by pali** [comment]
> I think you missed my point. Earlier I said "I could patch the NAT code or create firewall rules to fix the problem, but that's inconvenient, it's the same reason why people use miniupnpd instead of manually creating firewall rules."
> 
> "Patching NAT code" is patching the kernel. Alternatively firewall rules could be created every time client IP:port changes or new clients are added. Neither is convenient.

Look, I do not know why you to patch kernel code. And writing iptables/ipset rules is smaller then starting patching miniupnpd code, wait until new version is released and propagated to all distributions.

> As mentioned earlier, it's due to laziness... It's really the same reason why miniupnpd exists, people don't have to use miniupnpd as everything miniupnpd does can also be done manually instead.

miniupnpd is there to support IGD UPnP protocol. Something which cannot be implemented in netfiler, iptables or other NAT layer. Just because it is whole client<-->server general-purpose protocol which configure **explicit** port forwarding for **specified** period of time. Everything for which client ask. It is not bound to any technology or other network protocol on which connection tracking can be implemented -- just because it is for general purpose.

> Regarding the lifetime of the port-forwarding rules, it's a minor issue. NAT itself has a similar characteristic that applications cannot remove UDP translations or set the translation expiry, but we normally don't have problems with that. It shouldn't be difficult to deal with port-forwarding rule expiry. One solution is rely on the existing option "clean_ruleset_interval"

This is not minor issue. Timeouts in connection trackings or NAT traversal or similar techniques which open connection through NAT uses timings and keep alive packets. They are statefull and once NAT timeout occurs connections are closed. On the other hand, for port forwarding there are explicit stateless rules. Connection can be opened, closed, reconnected at any time, without need for any keep alive packets, without any NAT traversal technique. But once port forwarding rule is removed, it drop also untracked opened connections. So to have reliable connections, miniupnpd cannot remove port forwarding rule until connections are not closed. So miniupnpd would need to be extended to start doing some network traffic inspection or connection tracking. Because STUN protocol does not provide way to inform server that connection was closed.

> https://people.netfilter.org/chentschel/docs/sip-conntrack-nat.html

Conntrack helper modules can be written also as userspace applications. In mode when it just monitor SIP packets (without modifications) and just mark packets as RELATED, it should not produce new problems as packet marking is safe operation.

In similar way could be written userspace application which monitors those STUN binding packets and mark new incoming connections as RELATED.

But again, why to write tons of C code for something which is IIRC already possible to write via iptables and ipset? Because of laziness? It is the most stupid answer :-(

In the same way, I can ask question: Why not to integrate whole SIP forwarded into miniupnpd? It can act as SIP proxy, automatically adds/removes ip forwarding rules, based on SIP packets it can monitor when call starts, when finish, do all proper timings and correctly enable forwarding of RTP ports. Port forwarding is done in miniupnpd, so such SIP forwarded can be implemented in miniupnpd and via config options users can enable or disable them. And it is in miniupnpd because of laziness to implement it on other application layer where it fit better.

**Comment by miniupnp** [comment]
fixed by PR #307 / #314
